lib_eio/workpool.ml

```diff
@@ -0,0 +1,145 @@
+         to start waiting on the Stream before the current thread blocks on [Stream.take] again. *)
+      run_job job w;
+      while !capacity = 0 do
+        Condition.await_no_mutex condition
+      done;
+      (loop [@tailcall]) ()
+    | Quit { atomic; target; all_done } :: maybe_job ->
+      List.iter
+        (function
+          | Process job -> reject job
+          | _ -> assert false)
+        maybe_job;
+      (* Wait until the completion of all of this worker's jobs. *)
+      while !capacity < limit do
+        Condition.await_no_mutex condition
+      done;
+      (* If we're the last worker terminating, resolve the promise. *)
+      if Atomic.fetch_and_add atomic 1 = target then Promise.resolve all_done ()
+    | _ -> assert false
+  in
+  loop ()
+
+(* Start a new domain. The worker will need a switch, then we start the worker. *)
+let start_domain ~sw ~domain_mgr ~limit ~terminating ~transient stream =
+  let go () =
+    Domain_manager.run domain_mgr (fun () ->
+        Switch.run @@ fun sw -> start_worker ~sw ~limit ~terminating stream )
+  in
+  (* [transient] workpools run as daemons to not hold the user's switch from completing.
+     It's up to the user to hold the switch open (and thus, the workpool)
+     by blocking on the jobs issued to the workpool.
+     [Workpool.run] and [Workpool.run_exn] will block so this shouldn't be a problem.
+     Still, the user can call [Workpool.create] with [~transient:false] to
+     disable this behavior, in which case the user must call [Workpool.terminate]
+     to release the switch. *)
+  match transient with
+  | false -> Fiber.fork ~sw go
+  | true ->
+    Fiber.fork_daemon ~sw (fun () ->
+        go ();
+        `Stop_daemon )
+
+let create ~sw ~domain_count ~domain_concurrency ?(transient = true) domain_mgr =
+  let stream = Stream.create 0 in
+  let instance =
+    {
+      stream;
+      domain_count;
+      is_terminating = Atomic.make false;
+      terminating = Promise.create ();
+      terminated = Promise.create ();
+    }
+  in
+  let terminating = fst instance.terminating in
+  for _ = 1 to domain_count do
+    start_domain ~sw ~domain_mgr ~limit:domain_concurrency ~terminating ~transient stream
+  done;
+  instance
+
+let run_promise ~sw { stream; _ } f =
+  let p, w = Promise.create () in
+  Fiber.fork_promise ~sw (fun () ->
+      Stream.add stream (Process (Pack (f, w)));
+      Promise.await_exn p )
+
+let run { stream; _ } f =
+  let p, w = Promise.create () in
+  Stream.add stream (Process (Pack (f, w)));
+  Promise.await p
+
+let run_exn instance f =
+  match run instance f with
+  | Ok x -> x
+  | Error exn -> raise exn
+
+let terminate ~sw ({ terminating = _, w1; terminated = p2, w2; _ } as instance) =
+  if Atomic.compare_and_set instance.is_terminating false true
+  then (
+    (* Instruct workers to shutdown *)
+    Promise.resolve w1 (Quit { atomic = Atomic.make 1; target = instance.domain_count; all_done = w2 });
```

Comment:
This seems a bit over-complicated. I was expecting it to write `instance.domain_count` quit messages to the stream. That would avoid the need for a second channel for the quit message and the `n_any`.

Reply:
The reason it's this way is due to a combination of factors:
- I don't want the Quit messages to have to wait for the other enqueued jobs to run and complete ahead of them
- As much as possible, calling `terminate` should "immediately" halt starting new jobs.
The obvious solution is to make `terminate` reject all queued jobs before enqueueing Quit messages, but that doesn't work well with the post-termination background rejection loop, because `terminate` becomes both a producer and a consumer while the workers are still consumers. It can be made to work, but the end result was more complex and less predictable.

Reply:
> I don't want the Quit messages to have to wait for the other enqueued jobs to run and complete ahead of them.

The workers can still check `is_terminating` before running a job and reject the job if it's set. I think that would have the same effect.

Reply:
After trying it out, I'm remembering why it's not that way.
It's minor but I think it makes a difference.

By using a second channel, we're able to start the background rejection loop immediately. Otherwise, if all workers are fully occupied at the time `terminate` is called, we have to wait for a worker to be available to start rejecting jobs.

An alternative I've also explored is to immediately (T0) reject all queued jobs before enqueueing `n` Quit messages (T1), but that leaves jobs enqueued between T0 and T1 to hang until the background rejection job starts (which can only happen after all workers have exited so their Quit messages don't get dropped). This inconsistent behavior can be patched over by checking the `is_terminating` (`bool Atomic.t`) when submitting a new job but I'm trying to avoid all unnecessary thread coordination in the hot path...

Reply:
I think we have to do that anyway, otherwise this could happen:
1. Client checks `is_terminating` and see it's still OK.
2. Pool gets terminated; all workers finish; switch ends.
3. Client adds job to queue.

It's not clear what behaviour we really want for terminate though. e.g. why do we want to reject jobs that were submitted before `terminate` was called? Then there's no way for a client to behave "correctly" (so that it's jobs never get rejected) in the graceful shutdown case.

It would probably be helpful to have an example of a program that needs support for this kind of graceful shutdown (rather than just finishing the switch, which I suspect will cover most real uses). Or possibly we should make an initial version without `terminate` and add it later if/when it's needed.